### ISSUE TITLE
SIGSEGV caused on Android at "NativeMethods.sqlite3_close"

### ISSUE THREAD

**Issue by gentledepp** [ericsink/SQLitePCL.raw#137]
Hi!
We recently updated to the newest SQLite.net-PCL library (1.2.1) which - under the hoods - uses SQLitePCLRaw.bundle_green  version 1.1.0.

Now every once in a while, our app just crashes and in the logcat, we can see a SIGSEGV.
We are able to reproduce this issue on the Visual Studio Emulator for Android in the 10.1" KitKat (4.4) XHDPI Tablet API Lebel 19

the error looks like this:

    E/mono-rt (17852): Stacktrace:
    E/mono-rt (17852): 
    E/mono-rt (17852):   at <unknown> <0xffffffff>
    E/mono-rt (17852):   at (wrapper managed-to-native) SQLitePCL.SQLite3Provider_esqlite3/NativeMethods.sqlite3_close (intptr) <0x00012>
    E/mono-rt (17852):   at SQLitePCL.SQLite3Provider_esqlite3.SQLitePCL.ISQLite3Provider.sqlite3_close (intptr) <IL 0x00001, 0x0003f>
    E/mono-rt (17852):   at SQLitePCL.raw.sqlite3_close (SQLitePCL.sqlite3) <IL 0x0000b, 0x0007b>
    E/mono-rt (17852):   at SQLite.SQLite3.Close (SQLitePCL.sqlite3) [0x00000] in /Users/fak/Dropbox/Projects/sqlite-net/src/SQLite.cs:3300
    E/mono-rt (17852):   at SQLite.SQLiteConnection.Dispose (bool) [0x0006f] in /Users/fak/Dropbox/Projects/sqlite-net/src/SQLite.cs:1603
    E/mono-rt (17852):   at SQLite.SQLiteConnection.Close () [0x00000] in /Users/fak/Dropbox/Projects/sqlite-net/src/SQLite.cs:1589
    E/mono-rt (17852):   at Our.App.Core.Services.DataServiceBase.DisposeConnection (SQLite.SQLiteConnection) [0x00001] in C:\Code\icl\iCL\Our.App..Core\Services\DataServiceBase.cs:212
    E/mono-rt (17852):   at Our.App.Core.Services.DataServiceBase.UsingConnection (System.Action`1<SQLite.SQLiteConnection>) [0x0006a] in C:\Code\icl\iCL\Our.App..Core\Services\DataServiceBase.cs:122
    E/mono-rt (17852):   at Our.App.Core.Services.WorkbookDataService.Update (Our.App..Core.Entities.Workbooks.InspectionInfo) [0x00015] in C:\Code\icl\iCL\Our.App..Core\Services\WorkbookDataService.cs:743
    E/mono-rt (17852):   at Our.App.Core.Services.InspectionDataService.<UpdateInspectionProgress>b__104_0 (SQLite.SQLiteConnection) [0x00075] in C:\Code\icl\iCL\Our.App..Core\Services\InspectionDataService.cs:1372
    E/mono-rt (17852):   at Our.App.Core.Services.DataServiceBase/<>c__DisplayClass21_1`1<bool>.<UsingConnection>b__0 () [0x00000] in C:\Code\icl\iCL\Our.App..Core\Services\DataServiceBase.cs:144
    E/mono-rt (17852):   at Our.App.Core.Services.DataServiceBase.RetryOnConnectionError<bool> (System.Func`1<bool>) [0x00007] in C:\Code\icl\iCL\Our.App..Core\Services\DataServiceBase.cs:227
    E/mono-rt (17852):   at Our.App.Core.Services.DataServiceBase.UsingConnection<bool> (System.Func`2<SQLite.SQLiteConnection, bool>) [0x0004a] in C:\Code\icl\iCL\Our.App..Core\Services\DataServiceBase.cs:144
    E/mono-rt (17852):   at Our.App.Core.Services.InspectionDataService.UpdateInspectionProgress () [0x00001] in C:\Code\icl\iCL\Our.App..Core\Services\InspectionDataService.cs:1357
    E/mono-rt (17852):   at Our.App.Core.ViewModels.InspectionViewModel/<Init>d__173.MoveNext () [0x0053e] in C:\Code\icl\iCL\Our.App..Core\ViewModels\Inspection\InspectionViewModel.cs:650
    E/mono-rt (17852):   at System.Runtime.CompilerServices.AsyncMethodBuilderCore/MoveNextRunner.InvokeMoveNext (object) [0x00000] in /Users/builder/data/lanes/3511/501e63ce/source/mono/mcs/class/referencesource/mscorlib/system/runtime/compilerservices/AsyncMethodBuilder.cs:1089
    E/mono-rt (17852):   at System.Threading.ExecutionContext.RunInternal (System.Threading.ExecutionContext,System.Threading.ContextCallback,object,bool) [0x0008d] in /Users/builder/data/lanes/3511/501e63ce/source/mono/mcs/class/referencesource/mscorlib/system/threading/executioncontext.cs:957
    E/mono-rt (17852):   at System.Threading.ExecutionContext.Run (System.Threading.ExecutionContext,System.Threading.ContextCallback,object,bool) [0x00000] in /Users/builder/data/lanes/3511/501e63ce/source/mono/mcs/class/referencesource/mscorlib/system/threading/executioncontext.cs:904
    E/mono-rt (17852):   at System.Runtime.CompilerServices.AsyncMethodBuilderCore/MoveNextRunner.Run () [0x0003b] in /Users/builder/data/lanes/3511/501e63ce/source/mono/mcs/class/referencesource/mscorlib/system/runtime/compilerservices/AsyncMethodBuilder.cs:1070
    E/mono-rt (17852):   at System.Threading.Tasks.SynchronizationContextAwaitTaskContinuation.<s_postCallback>m__0 (object) [0x00000] in /Users/builder/data/lanes/3511/501e63ce/source/mono/mcs/class/referencesource/mscorlib/system/threading/Tasks/TaskContinuation.cs:388
    E/mono-rt (17852):   at Android.App.SyncContext/<Post>c__AnonStorey0.<>m__0 () [0x00000] in /Users/builder/data/lanes/3511/501e63ce/source/xamarin-android/src/Mono.Android/Android.App/SyncContext.cs:18
    E/mono-rt (17852):   at Java.Lang.Thread/RunnableImplementor.Run () [0x0000b] in /Users/builder/data/lanes/3511/501e63ce/source/xamarin-android/src/Mono.Android/Java.Lang/Thread.cs:36
    E/mono-rt (17852):   at Java.Lang.IRunnableInvoker.n_Run (intptr,intptr) [0x00009] in /Users/builder/data/lanes/3511/501e63ce/source/monodroid/src/Mono.Android/platforms/android-24/src/generated/Java.Lang.IRunnable.cs:81
    E/mono-rt (17852):   at (wrapper dynamic-method) object.036823a2-fea2-4fb9-b7c8-0c044f7e49eb (intptr,intptr) <IL 0x00011, 0x0001f>
    E/mono-rt (17852):   at (wrapper native-to-managed) object.036823a2-fea2-4fb9-b7c8-0c044f7e49eb (intptr,intptr) <IL 0x00021, 0x00097>
    E/mono-rt (17852): 
    E/mono-rt (17852): Attempting native Android stacktrace:
    E/mono-rt (17852): 
    E/mono-rt (17852): 
    E/mono-rt (17852): =================================================================
    E/mono-rt (17852): Got a SIGSEGV while executing native code. This usually indicates
    E/mono-rt (17852): a fatal error in the mono runtime or one of the native libraries 
    E/mono-rt (17852): used by your application.
    E/mono-rt (17852): =================================================================
    E/mono-rt (17852): 
    F/libc    (17852): Fatal signal 11 (SIGSEGV) at 0x3f800008 (code=1), thread 17852 (Our.App.)

At first we thought it might be because, we called ".Close()" and ".Dispose()" on the SQLite.SQLiteConnection, but removing the call to ".Close()" did not solve the issue for us.

Any thoughts?

**Comment by ericsink** [comment]
Hmmm.  So far I don't see anything that points me at a diagnosis.

You say you "recently updated".  I assume that this problem did not appear before that update.  What versions were you using before?

**Comment by gentledepp** [comment]
Before we were using MvvmCross.Community.Plugin.Sqlite version 3.1.1 which was using SQLite-net under the hood.
But it is really, really old - it was published 2014-02-09

https://www.nuget.org/packages/MvvmCross.Community.Plugin.Sqlite/3.1.1

**Comment by ericsink** [comment]
So you have effectively been using the system SQLite, which for API 19 was SQLite 3.7.11.  Now you're using a SQLite library build that is much more recent.

If I had to make a wild guess, I would say that your app is probably doing something SQLite doesn't like, and the older version was letting you get away with it, but the more recent version is not.

Is the sqlite connection handle valid?

Calling sqlite3_close() when there are prepared statements or blob handles that have not been released?

You might want to update your SQLitePCLRaw.bundle_green to the latest version, just in case it might make a difference, but I don't think it's likely.

**Comment by gentledepp** [comment]
well... the connection is used by sqlite-net pcl, so I doubt that there is any possibility that we could do anything wrong here like not releasing statements or the like.

What I find interesting, is that these random crashes happen on our 10" KitKat emulator the most, while on the 5" KitKat emulator we cannot even reproduce this issue. (We are running a battery of Xamarin.UITests on them)

I will try to run our tests on other devices as well and report back here. 
Thank you for your thoughts!

**Comment by gentledepp** [comment]
Could this be a threading issue?
What is the default mode of SQLitePCL.raw? "Serialized", "Multi-thread" or "Single-thread"?

**Comment by ericsink** [comment]
I suppose it could be a threading issue, although the fact that it only happens on one of your two otherwise identical emulators doesn't *seem* to support that notion.

The settings used to compile the e_sqlite3 builds for bundle_green are in:

https://github.com/ericsink/SQLitePCL.raw/blob/master/android/sqlite3/jni/Android.mk

The SQLITE_THREADSAFE compile option is not specified, and according to:

https://www.sqlite.org/threadsafe.html

the default is "serialized".

**Comment by gentledepp** [comment]
Update: I ran our UI tests on other devices as well (not just the emulators) and the crashes kept on coming.
However, once I updated to SQLitePCLRaw.bundle_green 1.1.2 I only got the error on an emulator once (insetad of in every of our 20 ui tests) and it did not happen on the native/real devices anymore.

Moreover, I noticed that the Visual Studio Emulators for Android are somewhat instable in general: I got random crashes on our UI tests that never happened on any device at all.

So thank you so much for your tip! (Can you think of any specific change in 1.1.2 that could have solved this?)

**Comment by ericsink** [comment]
The latest release updated the e_sqlite3 builds to 3.16.1. That might have made a difference in your situation. I'd have to review the change log for other changes to see if we could find any other likely possibilities.

**Comment by gentledepp** [comment]
never mind - I'll ping back though if those issues resurface again.
Thank you for the support and the great work you're doing! 👍

**Comment by gentledepp** [comment]
ok... so this is happening again all the time.
My hope: can you point me into the direction into how to debug this? (like the native side of SQLite?)

Because, I do not think that we're misusing the library, otherwise it should throw a more meaningful exception like "busy" or "invalid" or whatever at us, right?

**Comment by ericsink** [comment]
It is not necessarily true that misuse of the library would yield a helpful error code.

That said, I don't have any strong evidence to support the notion that the library is being misused.

Sorry for the bad timing, but I will be largely unavailable for the next two days due to travel.   I wish I could offer something more helpful.

**Comment by gentledepp** [comment]
Hi - me again!
We changed the call to the SQLiteNet.Connection from "connection.Dispose" to "connection.Close".
This causes the app to crash much less frequent, however, there are still crashes, like before:
`0 | 20.02.2017 13:08:46 | 7447 | 7447 | mono-rt Stacktrace:
0 | 20.02.2017 13:08:46 | 7447 | 7447 | mono-rt   at <unknown> <0xffffffff>
0 | 20.02.2017 13:08:46 | 7447 | 7447 | mono-rt   at (wrapper managed-to-native) SQLitePCL.SQLite3Provider_e_sqlite3/NativeMethods.sqlite3_close (intptr) <0x00012>
0 | 20.02.2017 13:08:46 | 7447 | 7447 | mono-rt   at SQLitePCL.SQLite3Provider_e_sqlite3.SQLitePCL.ISQLite3Provider.sqlite3_close (intptr) <0x00013>
0 | 20.02.2017 13:08:46 | 7447 | 7447 | mono-rt   at SQLitePCL.raw.sqlite3_close (SQLitePCL.sqlite3) <0x00025>
0 | 20.02.2017 13:08:46 | 7447 | 7447 | mono-rt   at SQLite.SQLite3.Close (SQLitePCL.sqlite3) <0x00013>
0 | 20.02.2017 13:08:46 | 7447 | 7447 | mono-rt   at SQLite.SQLiteConnection.Dispose (bool) <0x000df>
0 | 20.02.2017 13:08:46 | 7447 | 7447 | mono-rt   at SQLite.SQLiteConnection.Close () <0x00018>
0 | 20.02.2017 13:08:46 | 7447 | 7447 | mono-rt   at My.SuperDuperApp.Core.Services.DataServiceBase.DisposeConnection (SQLite.SQLiteConnection) <0x00013>
0 | 20.02.2017 13:08:46 | 7447 | 7447 | mono-rt   at My.SuperDuperApp.Core.Services.DataServiceBase.UsingConnection (System.Action`1<SQLite.SQLiteConnection>) <0x001c7>
0 | 20.02.2017 13:08:46 | 7447 | 7447 | mono-rt   at My.SuperDuperApp.Core.Services.WorkbookDataService.Update (My.SuperDuperApp.Core.Entities.Workbooks.InspectionInfo) <0x000df>
0 | 20.02.2017 13:08:46 | 7447 | 7447 | mono-rt   at My.SuperDuperApp.Core.Services.InspectionDataService.<UpdateInspectionProgress>b__104_0 (SQLite.SQLiteConnection) <0x00164>
0 | 20.02.2017 13:08:46 | 7447 | 7447 | mono-rt   at My.SuperDuperApp.Core.Services.DataServiceBase/<>c__DisplayClass22_1`1<bool>.<UsingConnection>b__0 () <0x0001e>
0 | 20.02.2017 13:08:46 | 7447 | 7447 | mono-rt   at My.SuperDuperApp.Core.Services.DataServiceBase.RetryOnConnectionError<bool> (System.Func`1<bool>) <0x0003f>
0 | 20.02.2017 13:08:46 | 7447 | 7447 | mono-rt   at My.SuperDuperApp.Core.Services.DataServiceBase.UsingConnection<bool> (System.Func`2<SQLite.SQLiteConnection, bool>) <0x0018f>
0 | 20.02.2017 13:08:46 | 7447 | 7447 | mono-rt   at My.SuperDuperApp.Core.Services.InspectionDataService.UpdateInspectionProgress () <0x00077>
0 | 20.02.2017 13:08:46 | 7447 | 7447 | mono-rt   at My.SuperDuperApp.Core.ViewModels.InspectionViewModel/<Init>d__173.MoveNext () <0x00ebc>
0 | 20.02.2017 13:08:46 | 7447 | 7447 | mono-rt   at System.Runtime.CompilerServices.AsyncMethodBuilderCore/MoveNextRunner.InvokeMoveNext (object) <0x00082>
0 | 20.02.2017 13:08:46 | 7447 | 7447 | mono-rt   at System.Threading.ExecutionContext.RunInternal (System.Threading.ExecutionContext,System.Threading.ContextCallback,object,bool) <0x001b6>
0 | 20.02.2017 13:08:46 | 7447 | 7447 | mono-rt   at System.Threading.ExecutionContext.Run (System.Threading.ExecutionContext,System.Threading.ContextCallback,object,bool) <0x00027>
0 | 20.02.2017 13:08:46 | 7447 | 7447 | mono-rt   at System.Runtime.CompilerServices.AsyncMethodBuilderCore/MoveNextRunner.Run () <0x0009f>
0 | 20.02.2017 13:08:46 | 7447 | 7447 | mono-rt   at System.Threading.Tasks.SynchronizationContextAwaitTaskContinuation.<s_postCallback>m__0 (object) <0x00024>
0 | 20.02.2017 13:08:46 | 7447 | 7447 | mono-rt   at Android.App.SyncContext/<Post>c__AnonStorey0.<>m__0 () <0x0001b>
0 | 20.02.2017 13:08:46 | 7447 | 7447 | mono-rt   at Java.Lang.Thread/RunnableImplementor.Run () <0x00027>
0 | 20.02.2017 13:08:46 | 7447 | 7447 | mono-rt   at Java.Lang.IRunnableInvoker.n_Run (intptr,intptr) <0x00034>
0 | 20.02.2017 13:08:46 | 7447 | 7447 | mono-rt   at (wrapper dynamic-method) object.1cd1cc87-b35d-472d-9cd8-8573c1aba11c (intptr,intptr) <0x0001f>
0 | 20.02.2017 13:08:46 | 7447 | 7447 | mono-rt   at (wrapper native-to-managed) object.1cd1cc87-b35d-472d-9cd8-8573c1aba11c (intptr,intptr) <0x00073>
 `

unfortunately, this is quite annoying, as our monitoring system (Microsoft Mobile Center) does not seem to track these (native SIGSEGV) exceptions.

Can you give ma any thoughts in how I could possibly debug this?

Also: We have a repository base class which does all the connection handling.
Basically, we use a single connection for each call to the database. To share the same connection through the logical async call stack, it uses "CallContext.LogicalSetData" and "CallContext.LogicalGetData"
Could that be the issue? :-|

**Comment by ericsink** [comment]
"Could that be the issue?"

I can't be sure, but it sounds suspicious.

Is there any chance your code is trying to close the connection twice?

**Comment by gentledepp** [comment]
I am very shure about that, but I'll double check.
Anyways, I would expect some meaningful exception in that case, or is that too much to expect? :)

Would it be possible to find out, whether a SQLiteConnection has already been closed? (like "IsClosed" or whatever?)

Thanks for the input!

**Comment by ericsink** [comment]
That was just a random guess on my part.  I'm not even sure that a double-close would cause the symptom you are seeing.

Your code seems to be doing *something* that the native sqlite lib does not like, because the crash is down there in native code.  That's why you get such a non-helpful exception message.  So I'm just trying to guess what is going on.

AFAIK, no, the sqlite API does not provide a way to tell if a connection handle has already been closed.

**Comment by gentledepp** [comment]
hmmm... so I took a look at this and it has definetly nothing to do with disposing it twice.

1. If we call `conn.Close(); ` two times in a row, the second call is only ignored

2. Additionally, we noticed that the "conn.Handle" is set to null after the first call to Close, so to make sure, that the method is not called concurrently, we added this lock statement
    private void DisposeConnection(SQLiteConnection conn)
    {
        if (conn.Handle != null)
        {
            lock (conn.Handle)
            {
                if(conn.Handle != null)
                {
                    conn.Close(); 
                }
            }
        }
    }


However, that had absolutely no effect, so I expect there is a bug in SQLite then.... 👎
